### ISSUE TITLE
Data mismatch when transferring data from Oracle DB to Elastic Search

### ISSUE THREAD

**Issue by Sarvesh-D** [confluentinc/kafka-connect-jdbc#256]
Hi, we are using kafka JDBC connect to select data from oracle database and pump it into elastic search using elastic search sink. We are observing that there is ambiguity in data in DB and Elastic search. Below is a single record as it looks in DB table and in elastic search:

Table description:
```
Name                  Null     Type               
--------------------- -------- ------------------ 
ID                   NOT NULL NUMBER(19)         
ID_1         NOT NULL VARCHAR2(255 CHAR) 
NAME       NOT NULL VARCHAR2(255 CHAR) 
TYPE_ID             NUMBER(19)         
CREATED_BY            NOT NULL VARCHAR2(100)      
```

DB export of single record as JSON:
```
{
  "items": [
    {
      "id": 2,
      "id_1": "TEST",
      "name": "",
      "type_id": 1,
      "created_by": "Batch"
    }
  ]
}
```

Same record in Elastic Search:
```
{
  "_index": "es-test",
  "_type": "kafka-connect",
  "_id": "nfs-jdbc-TEST+0+2",
  "_score": 11.047933,
  "_source": {
    "ID": "Ag==",
    "ID_1": "TEST",
    "NAME": ""
    "TYPE_ID": "AQ=="
    "CREATED_BY": "Batch"
  }
}
```

If you see there is difference in below fields as observed in DB and ES:

1. ID
2. TYPE_ID

**Note that both these fields are of type NUMBER.**

The connect-avro-standalone .properties read as below:
```
bootstrap.servers=localhost:9092
key.converter=io.confluent.connect.avro.AvroConverter
key.converter.schema.registry.url=http://localhost:8082
value.converter=io.confluent.connect.avro.AvroConverter
value.converter.schema.registry.url=http://localhost:8082
internal.key.converter=org.apache.kafka.connect.json.JsonConverter
internal.value.converter=org.apache.kafka.connect.json.JsonConverter
internal.key.converter.schemas.enable=false
internal.value.converter.schemas.enable=false
offset.storage.file.filename=/tmp/connect.offsets
```

P.S: We also changed internal.key.converter and internal.value.converter to AvroConverter, still the same issue.

**Comment by ewencp** [comment]
@Sarvesh-D This has to do with the fact that Number columns can have precisions that are not safe to represent as integers or longs so they default to using the Decimal logical type which allows for arbitrary precision and is encoded as bytes (which in JSON are getting converted to base64-encoded strings). There's a long discussion of different options to improve this behavior [here](https://github.com/confluentinc/kafka-connect-jdbc/issues/101). Another alternative would be to use a [single message transform](http://kafka.apache.org/documentation.html#connect_transforms) to coerce the type.

**Comment by Sarvesh-D** [comment]
@ewencp Thanks for the suggestion. I tried to use Cast transformer as a way to coerce the type. Below is the configuration i added to kafka-connect-jdbc/source-quickstart-sqlite.properties file:
```
transforms=CastNumbers
transforms.CastNumbers.type=org.apache.kafka.connect.transforms.Cast$Value
transforms.CastNumbers.spec=ID:int32
```
However, while running the connect i am getting below exception:
```
java.util.concurrent.ExecutionException: org.apache.kafka.connect.runtime.rest.errors.BadRequestException: Connector configuration is invalid and contains the following 2 error(s):
Invalid value org.apache.kafka.connect.transforms.Cast$Value for configuration transforms.CastNumbers.type: Class org.apache.kafka.connect.transforms.Cast$Value could not be found.
Invalid value null for configuration transforms.CastNumbers.type: Not a Transformation
You can also find the above list of errors at the endpoint `/{connectorType}/config/validate`
	at org.apache.kafka.connect.util.ConvertingFutureCallback.result(ConvertingFutureCallback.java:80)
	at org.apache.kafka.connect.util.ConvertingFutureCallback.get(ConvertingFutureCallback.java:67)
	at org.apache.kafka.connect.cli.ConnectStandalone.main(ConnectStandalone.java:96)
Caused by: org.apache.kafka.connect.runtime.rest.errors.BadRequestException: Connector configuration is invalid and contains the following 2 error(s):
Invalid value org.apache.kafka.connect.transforms.Cast$Value for configuration transforms.CastNumbers.type: Class org.apache.kafka.connect.transforms.Cast$Value could not be found.
Invalid value null for configuration transforms.CastNumbers.type: Not a Transformation
You can also find the above list of errors at the endpoint `/{connectorType}/config/validate`
	at org.apache.kafka.connect.runtime.AbstractHerder.maybeAddConfigErrors(AbstractHerder.java:365)
	at org.apache.kafka.connect.runtime.standalone.StandaloneHerder.putConnectorConfig(StandaloneHerder.java:158)
	at org.apache.kafka.connect.cli.ConnectStandalone.main(ConnectStandalone.java:93)
```
I am using confluent-3.2.2. Do i need to download some other binaries?

**Comment by Sarvesh-D** [comment]
@ewencp please help me configure the transformer, i am still not able to configure it.

**Comment by rrangine** [comment]
@Sarvesh-D Any luck with this approach? I used the below :+1: 
"transforms":"CastNumbers",
"transforms.CastNumbers.type":"org.apache.kafka.connect.transforms.Cast$Value",
"transforms.CastNumbers.spec":"PRIORITY:string"

Where PRIORITY is type: NUMBER.  The transformation threw below exception on Connect:

org.apache.kafka.connect.errors.DataException: Unexpected type in Cast transformation: BYTES
        at org.apache.kafka.connect.transforms.Cast.convertFieldType(Cast.java:207)
        at org.apache.kafka.connect.transforms.Cast.getOrBuildSchema(Cast.java:168)
        at org.apache.kafka.connect.transforms.Cast.applyWithSchema(Cast.java:138)
        at org.apache.kafka.connect.transforms.Cast.apply(Cast.java:108)
        at org.apache.kafka.connect.runtime.TransformationChain.apply(TransformationChain.java:38)
        at org.apache.kafka.connect.runtime.WorkerSourceTask.sendRecords(WorkerSourceTask.java:190)
        at org.apache.kafka.connect.runtime.WorkerSourceTask.execute(WorkerSourceTask.java:168)
        at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:146)
        at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:190)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
@ewencp  @astubbs Any thoughts?

**Comment by satishsg** [comment]
@ewencp I am using the jdbc connector and have the same issue as @Sarvesh-D with casting. can you please let me know how to resolve it?
`[2017-09-12 10:21:28,204] ERROR Stopping after connector error (org.apache.kafka.connect.cli.ConnectStandalone:99)
java.util.concurrent.ExecutionException: org.apache.kafka.connect.runtime.rest.errors.BadRequestException: Connector configuration is invalid and contains the following 2 error(s):
Invalid value org.apache.kafka.connect.transforms.Cast$Key for configuration transforms.castFields.type: Class org.apache.kafka.connect.transforms.Cast$Key could not be found.
Invalid value null for configuration transforms.castFields.type: Not a Transformation
You can also find the above list of errors at the endpoint `/{connectorType}/config/validate`
	at org.apache.kafka.connect.util.ConvertingFutureCallback.result(ConvertingFutureCallback.java:80)
	at org.apache.kafka.connect.util.ConvertingFutureCallback.get(ConvertingFutureCallback.java:67)
	at org.apache.kafka.connect.cli.ConnectStandalone.main(ConnectStandalone.java:96)
Caused by: org.apache.kafka.connect.runtime.rest.errors.BadRequestException: Connector configuration is invalid and contains the following 2 error(s):
Invalid value org.apache.kafka.connect.transforms.Cast$Key for configuration transforms.castFields.type: Class org.apache.kafka.connect.transforms.Cast$Key could not be found.
Invalid value null for configuration transforms.castFields.type: Not a Transformation
You can also find the above list of errors at the endpoint `/{connectorType}/config/validate`
	at org.apache.kafka.connect.runtime.AbstractHerder.maybeAddConfigErrors(AbstractHerder.java:365)
	at org.apache.kafka.connect.runtime.standalone.StandaloneHerder.putConnectorConfig(StandaloneHerder.java:158)
	at org.apache.kafka.connect.cli.ConnectStandalone.main(ConnectStandalone.java:93)
`

**Comment by ewencp** [comment]
If you're getting

```
Invalid value org.apache.kafka.connect.transforms.Cast$Value for configuration transforms.CastNumbers.type: Class org.apache.kafka.connect.transforms.Cast$Value could not be found.
```

this means the transformation cannot be found -- either the jar isn't on your classpath or you're using too old a version. Connect only shipped with `Cast` as of 0.11.

@rrangine Your issue is that you're getting the Decimal logical type. Unfortunately support for that has not yet been added into the Cast transformation so you'd need to extend it to work -- I had hoped to get it in with the first version of that transformation but unfortunately I didn't have enough time.

**Comment by rmoff** [comment]
As of Confluent Platform 4.1.1 the way to handle this is with `numeric.precision` in the JDBC source config properties. See https://docs.confluent.io/current/connect/kafka-connect-jdbc/source-connector/index.html#mapping-column-types.
